content/blog/2020-11-16-reproducibility-in-ml-my-tools/index.en.md

```diff
@@ -0,0 +1,122 @@
+
+``` yaml
+data_loader:
+    batch_size: 2048
+
+training_settings:
+    seed: 42
+    device: "cuda:0"
+```
+Figure 3:  Example of a YAML file. [Here](https://stackoverflow.com/a/1729545/7927776) is a nice Stack Overflow answer about the difference between YAML and JSON files.
+
+
+Second, parameters need to be hierarchical. I only want to use the parameters for a specific case without having an obligation to have other parameters I don't need. For example, if a want to compare the performance using SGD and Adam optimizer, I will use two sets of parameters: a learning rate for SGD and a learning rate and beta values for Adam. If I'm using Argparse, I would need to have beta parameters even when using SGD.
+
+[Hydra](https://hydra.cc/) solves both by hierarchically using YAML files. With Hydra, you can compose your configuration dynamically, enabling you to easily get the configuration you want. That is, you can enable SGD settings (`optimizer: SGD`) or Adam settings (`optimizer: Adam`) (see Figure 4). This way of "calling" your configuration means that you will always get the settings you need and not those of other configurations. Also, this hierarchical configuration management is straightforward to understand, as shown in Figure 5. We can see four different models and can access their parameters easily. For sure, if you only have 2 or 3 parameters, it seems like overkill, but how long will you only have 2 or 3 parameters? I don't know about you, but not much longer after I start a project, I rapidly have more than 3.
+
+``` yaml
+data_loader:
+    batch_size: 2048
+
+training_settings:
+    seed: 42
+    device: "cuda:0"
+
+defaults:
+    - optimizer: SGD # call the SGD YAML file
+    - model: bi_lstm
+    - dataset: canadian
+    - embeddings: fast_text
+```
+Figure 4:  Example of a YAML file when you use hierarchical configuration. `optimizer: SGD` is equivalent to the content of the file `conf/optimizer/SGD.yaml` in the Figure 5.
+
+``` sh
+.
+├── config.yaml
+├── dataset
+│   ├── all.yaml
+│   └── canadian.yaml
+├── embeddings
+│   ├── fast_text_character.yaml
+│   └── fast_text.yaml
+├── model
+│   ├── bi_lstm_bidirectionnal.yaml
+│   ├── bi_lstm.yaml
+│   ├── lstm_bidirectionnal.yaml
+│   └── lstm.yaml
+└── optimizer
+    ├── adam.yaml
+    └── SGD.yaml
+```
+Figure 5: Example of a hierarchical configuration directory to rapidly manage your settings.
+
+## Conclusion
+Lack of reproducibility in your machine learning projects can lead to a considerable slowdown to put your models into production. I've presented two solutions to solve some of the problems in your machine learning project. These solutions will help you manage your configuration, experiments and your results. For a more complete presentation, see my [seminar](https://davebulaval.github.io/gestion-configuration-resultats/).
```

Comment:
Y'a un léger truc dans la 1ere phrase qui me fait douter: _lead to a considerable slowdown to put_ .  
```suggestion
Lack of reproducibility in your machine learning projects can lead to a considerable slowdown to put your models into production. I've presented two solutions for your machine learning projects. I hope that these will help you manage your configurations, experiments and your results. For a more complete presentation, see my [seminar](https://davebulaval.github.io/gestion-configuration-resultats/).
```